collated/main/A0121533W.md

```diff
@@ -0,0 +1,797 @@
+     * dateTimeSource
+     * 
+```
+###### \java\tars\commons\util\MarkTaskTracker.java
+``` java
+ *
+ */
+public class MarkTaskTracker {
+    
+    private ArrayList<Integer> markDoneTasks;
+    private ArrayList<Integer> markUndoneTasks;
+    private ArrayList<Integer> alreadyDoneTasks;
+    private ArrayList<Integer> alreadyUndoneTasks;
+    
+    public static final String SUCCESS_DONE = "Task: %1$s marked done successfully.\n"; 
+    public static final String SUCCESS_UNDONE = "Task: %1$s marked undone successfully.\n"; 
+    public static final String ALREADY_DONE = "Task: %1$s already marked done.\n"; 
+    public static final String ALREADY_UNDONE = "Task: %1$s already marked undone.\n"; 
+    
+    
+    /**
+     * Constructor
+     */
+    public MarkTaskTracker() {
+        this.markDoneTasks = new ArrayList<Integer>();
+        this.markUndoneTasks = new ArrayList<Integer>();
+        this.alreadyDoneTasks = new ArrayList<Integer>();
+        this.alreadyUndoneTasks = new ArrayList<Integer>();
+    }
+    
+    /**
+     * Adds target index of task to relevant "To Mark List" based on status
+     * @param targetIndex
+     * @param status Done or Undone
+     */
+    public void addToMark(int targetIndex, Status status) {
+        if (status.status) {
+            addToMarkDoneTask(targetIndex);
+        } else {
+            addToMarkUndoneTask(targetIndex);
+        }
+        
+    }
+
+    /**
+     * Adds target index of task to relevant "Already Marked List" based on status
+     * @param targetIndex
+     * @param status Done or Undone
+     */
+    public void addAlreadyMarked(int targetIndex, Status status) {
+        if (status.status) {
+            addToAlreadyDoneTasks(targetIndex);
+        } else {
+           addToAlreadyUndoneTasks(targetIndex);
+        }
+    }
+
+    /**
+     * Return string for each tasks index in the specific ArrayLists
+     * @return
+     */
+    public String getResult() {
+        String markDoneTasksString = getIndexesString(markDoneTasks);
+        String markUndoneTasksString = getIndexesString(markUndoneTasks);
+        String alreadyDoneTasksString = getIndexesString(alreadyDoneTasks);
+        String alreadyUndoneTasksString = getIndexesString(alreadyUndoneTasks);
+        
+        String result = formatResults(markDoneTasksString, markUndoneTasksString, 
+                alreadyDoneTasksString, alreadyUndoneTasksString);
+        
+        return result;
+    }
+
+    private String formatResults(String markDoneTasksString, String markUndoneTasksString,
+            String alreadyDoneTasksString, String alreadyUndoneTasksString) {
+        
+        String markDoneResult = "";
+        String markUndoneResult = "";
+        String alreadyDoneResult = "";
+        String aldreadyUndoneResult = "";
+        
+        if (markDoneTasksString.length() != 0) {
+            markDoneResult = String.format(SUCCESS_DONE, markDoneTasksString);
+        }
```

Comment:
These lines can be SLAP into a generic function , notice the similarity in their code format